### PR TITLE
[BACKPORT ] v3.12 Clear resources on start fail #690

### LINKED ISSUE #690: Client does not clean up resources on failed start (backport required)
Issue for tracking 3.12.z patch for #687

### PR DESCRIPTION
**Pull request by srknzl**
Hi @puzpuzpuz,

## Changes

* Clear resources on start failure
* Small changes to some tests that are done in other branches
*  /bin/sh made /bin/bash in scripts/download-rc.sh hashbang to support [[ and ]] signs 

fixes #690

### CODE REVIEW

File: test/ClientProxyTest.js [hunk 1]
```diff
@@ -52,9 +52,9 @@ describe('Generic proxy test', function () {
         assert.equal(mapProxy.getConnectedServerVersion(), -1);
     });
 
-    it('Client with a 3.7 server connection should return the version', function () {
```
Comment (puzpuzpuz) [inline test/ClientProxyTest.js:55]: nit: it makes sense to leave 3.7 version here as this branch is only compatible with IMDG 3.x.

**Review by puzpuzpuz — COMMENTED** [review]
Could you include a backport of [`ClientShutdownTest`](https://github.com/hazelcast/hazelcast-nodejs-client/pull/689/files#diff-da9a095a1f633123c682589e0c4001680f9761677b0df5b737b1e38cd2734514) into this PR?

**Comment by srknzl** [comment]
> Could you include a backport of [`ClientShutdownTest`](https://github.com/hazelcast/hazelcast-nodejs-client/pull/689/files#diff-da9a095a1f633123c682589e0c4001680f9761677b0df5b737b1e38cd2734514) into this PR?

@puzpuzpuz Added it I am ready for another review. Thanks :)

File: test/ClientShutdownTest.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    afterEach(async function () {
+        sandbox.restore();
+        if (cluster != null) {
+            await RC.terminateCluster(cluster.id);
+            cluster = null;
+        }
+    });
+
+    it('client should call shutdown on failed start', async function () {
```
Comment (puzpuzpuz) [inline test/ClientShutdownTest.js:51]: Node.js client v3.12.x is supposed to support Node.js 4+, so we can't use async/await syntax in tests. Could you update the test in Promise callbacks syntax?

File: test/ClientShutdownTest.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+        expect(RepairingTask.prototype.shutdown.calledOnce).to.be.true;
+        expect(NearCacheManager.prototype.destroyAllNearCaches.calledOnce).to.be.true;
+        expect(Statistics.prototype.stop.calledOnce).to.be.true;
+        expect(PartitionService.prototype.shutdown.calledOnce).to.be.true;
+        expect(Heartbeat.prototype.cancel.calledOnce).to.be.true;
+        expect(ClientConnectionManager.prototype.shutdown.calledOnce).to.be.true;
+        expect(ListenerService.prototype.shutdown.calledOnce).to.be.true;
+        expect(InvocationService.prototype.shutdown.calledOnce).to.be.true;
+
```
Comment (puzpuzpuz) [inline test/ClientShutdownTest.js:85]: nit: redundant newline here

**Comment by srknzl** [comment]
I wrote chai promise tests using then() and catch() first time, so If I did something wrong inform pls @puzpuzpuz thanks :)

File: test/ClientShutdownTest.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            }).catch(err => {
+                throw err;
+            });
+        }
+    });
+
+    it('client should call shutdown on failed start', function () {
+        sandbox.spy(Client.prototype, 'shutdown');
+
+        expect(Client.newHazelcastClient().catch(
```
Comment (puzpuzpuz) [inline test/ClientShutdownTest.js:57]: Something like the following should work better:
```js
        return expect(Client.newHazelcastClient())
            .to.be.rejectedWith(HzErrors.IllegalStateError)
            .then(() => {
                expect(Client.prototype.shutdown.calledOnce).to.be.true;
            });
```

Note: mocha expects the function to return a Promise (see the `return expect(...)` change).

File: test/ClientShutdownTest.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            () => {
+                expect(Client.prototype.shutdown.calledOnce).to.be.true;
+            }
+        )).to.be.rejectedWith(HzErrors.IllegalStateError);
+
```
Comment (puzpuzpuz) [inline test/ClientShutdownTest.js:62]: nit: redundant newline here

**Comment by srknzl** [comment]
Thanks, Done!

**Review by puzpuzpuz — APPROVED** [review]
LGTM. Thanks for the contribution!

**Comment by puzpuzpuz** [comment]
verify

File: test/ClientShutdownTest.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+var expect = chai.expect;
+var chaiAsPromised = require('chai-as-promised');
+chai.use(chaiAsPromised);
+
+var sinon = require('sinon');
+var sandbox = sinon.createSandbox();
+
+var RC = require('./RC');
+var HzErrors = require('../.').HazelcastErrors;
+var {Client} = require('../.');
```
Comment (puzpuzpuz) [inline test/ClientShutdownTest.js:29]: I got failed CI build as Node.js 4 doesn't support destructuring assignments. Could you replace them with direct assignments?

Comment (srknzl) [reply test/ClientShutdownTest.js:29]: Ok

Comment (srknzl) [reply test/ClientShutdownTest.js]: Ready

**Comment by puzpuzpuz** [comment]
verify

**Comment by puzpuzpuz** [comment]
verify

**Comment by puzpuzpuz** [comment]
CI fails with the following error:
```
Unhandled rejection AssertionError: expected undefined to be true
    at /home/puzpuzpuz/projects/hazelcast-nodejs-client/test/ClientShutdownTest.js:80:70
    at tryCatcher (/home/puzpuzpuz/projects/hazelcast-nodejs-client/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/home/puzpuzpuz/projects/hazelcast-nodejs-client/node_modules/bluebird/js/release/promise.js:547:31)
    at Promise._settlePromise (/home/puzpuzpuz/projects/hazelcast-nodejs-client/node_modules/bluebird/js/release/promise.js:604:18)
    at Promise._settlePromise0 (/home/puzpuzpuz/projects/hazelcast-nodejs-client/node_modules/bluebird/js/release/promise.js:649:10)
    at Promise._settlePromises (/home/puzpuzpuz/projects/hazelcast-nodejs-client/node_modules/bluebird/js/release/promise.js:729:18)
    at _drainQueueStep (/home/puzpuzpuz/projects/hazelcast-nodejs-client/node_modules/bluebird/js/release/async.js:93:12)
    at _drainQueue (/home/puzpuzpuz/projects/hazelcast-nodejs-client/node_modules/bluebird/js/release/async.js:86:9)
    at Async._drainQueues (/home/puzpuzpuz/projects/hazelcast-nodejs-client/node_modules/bluebird/js/release/async.js:102:5)
    at Immediate.Async.drainQueues [as _onImmediate] (/home/puzpuzpuz/projects/hazelcast-nodejs-client/node_modules/bluebird/js/release/async.js:15:14)
    at processImmediate [as _immediateCallback] (timers.js:396:17)
```

That's probably because a `return` is missing here: https://github.com/hazelcast/hazelcast-nodejs-client/pull/796/files#diff-da9a095a1f633123c682589e0c4001680f9761677b0df5b737b1e38cd2734514R72

**Comment by srknzl** [comment]
got it, you can add the change if you want.

**Comment by srknzl** [comment]
Ok added the return

**Comment by puzpuzpuz** [comment]
verify

**Comment by puzpuzpuz** [comment]
Hopefully, I've fixed `ClientShutdownTest`. See https://github.com/srknzl/hazelcast-nodejs-client/pull/1

**Comment by srknzl** [comment]
if you merge pls change commit message as I used your commit message accidentally :)

**Comment by srknzl** [comment]
we can try again. was the error due to this?

**Comment by puzpuzpuz** [comment]
verify

**Comment by puzpuzpuz** [comment]
@srknzl thanks for your contribution!

**Comment by srknzl** [comment]
Np, thanks for the help

**Pull request merged**